### DISCUSSION TITLE
Welcome to deeptime Discussions!

### DISCUSSION THREAD

**Discussion by clonker** [deeptime-ml/deeptime#108] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by tommysisk** [comment]
Hello, is there a simple way to estimate a ck-test for an msm WITHOUT doing it in the space of "meta-stable sets"? i.e I want to see what my ck tests looks like from the initial clustering/discrete trajectory with the SAME indices. By default, I get ck-tests re-indexed  by meta-stable set even if number of metastable states == number of initial states unless I manually set meta-stable assignment probabilities equal to the Identity matrix. Thanks!

**Reply by clonker** [reply]
Hi Tommy,
this is currently a bit hidden but well possible:

```python
import numpy as np

import matplotlib.pyplot as plt

from deeptime.clustering import KMeans
from deeptime.data import ellipsoids
from deeptime.markov._observables import MembershipsObservable
from deeptime.markov.msm import MaximumLikelihoodMSM
from deeptime.plots import plot_ck_test
from deeptime.util.validation import ck_test

dtraj = ellipsoids().discrete_trajectory(10000)
otraj = ellipsoids().map_discrete_to_observations(dtraj)

dtraj2 = KMeans(n_clusters=10).fit_transform(otraj)

models = []

for lag in [1, 2, 3, 4, 5, 6]:
    models.append(MaximumLikelihoodMSM(lagtime=1).fit(dtraj2).fetch_model())

observable = MembershipsObservable(test_model=models[0], memberships=np.eye(models[0].n_states))
test_result = ck_test(models=models, observable=observable, test_model=models[0])

plot_ck_test(test_result)
plt.show()
```

But please be aware that with a lot of states you also need a lot of data to achieve sufficient statistics for the CK test. This is why we typically look at it in coarse grained space, otherwise it is simply not feasible.

Cheers!

**Reply by tommysisk** [reply]
Thanks!!

**Comment by tommysisk** [comment]
Another question: Can I use the deeptime.utils.data.Trajectorydataset class to make a time lagged trajectory dataset which has multiple inputs to the network at each time-step. To elaborate a bit more.. I have built a mixed input nn module with torch that takes a matrix and two vectors of data at each iteration, running the matrix through conv layers and vectorized data through linear layers before concatenating the outputs and eventually spitting out a softmax distribution over the markov states (skipping over many details, here). Is there a data structure that I can use to hold multiple tenors per time-step that will work with deeptime.utils.data.Trajectorydataset? I have accomplished this by squeezing all of the data into a 4 dimensional tensor and then slicing each vector/matrix out of the input in the forward method but this does not scale well and makes for odd looking code.

**Reply by tommysisk** [reply]
Here's just the forward method to hopefully provide clarity

def forward(self,x):
        image = x[:,:,:-2,:]
        image = self.convs(image)
        image = F.relu(self.fc1(image))
        vector = x[:,:,-2:-1,:-6].reshape(-1,15)
        vector = self.numeric_features(vector)
        vector2 = x[:,:,-1:,:].reshape(-1,21)
        vector2 = self.numeric_features2(vector2)
        combined = torch.cat((image, vector,vector2), 1)
        combined = self.combined_features(combined)
        return combined

**Reply by clonker** [reply]
Not right now, no. But it should be relatively easy to do by hand. A trajectory DS contains nothing more but pairs X and Y of instantaneous and time-lagged data.
